mlserver/batching/requests.py

```diff
@@ -0,0 +1,121 @@
+from operator import mul
+from collections import defaultdict
+from typing import Any, Dict, Iterable, List, Union
+from functools import reduce
+
+from ..types import (
+    InferenceRequest,
+    InferenceResponse,
+    RequestInput,
+    ResponseOutput,
+)
+
+
+def _get_data(payload: Union[RequestInput, ResponseOutput]):
+    return getattr(payload.data, "__root__", payload.data)
+
+
+def _infer_elem_size(shape: List[int]) -> int:
+    # TODO: Allow to use a different batch dimension
+    return reduce(mul, shape[1:], 1)
+
+
+def _merge_data(request_inputs: List[RequestInput]) -> Any:
+    all_data = [_get_data(request_input) for request_input in request_inputs]
+
+    sampled_datum = all_data[0]
+
+    if isinstance(sampled_datum, str):
+        return "".join(all_data)
+
+    if isinstance(sampled_datum, bytes):
+        return b"".join(all_data)
+
+    if isinstance(sampled_datum, list):
+        return sum(all_data, [])
+
+    # TODO: Should we raise an error if we couldn't merge the data?
```

Comment:
Hmm good question - I think we should be explicit on how we deal with errors, namely whether we'd rather ensure "it still gets processed" even if has to be sent without adaptive batching...

Reply:
I'd say raise an error - batches need to be same shape - so better to be explicit than try to predict.

Reply:
I was thinking of falling back to non-batched prediction, which shouldn't be too tricky. However, this is more of an argument between being explicit or not. 

I guess there's a strong argument for raising that to the user so that they're aware that the server couldn't batch a particular request. Although we could also make that clear through logging maybe?

Reply:
But is it valid for this to happen if so its not an error.